payments/controllers_test.go

```diff
@@ -0,0 +1,177 @@
+	if err != nil {
+		t.Error("failed to init service: ", err)
+	}
+
+	secretsStored := false
+	confStored := false
+
+	r := chi.NewRouter()
+	// startup our configuration middleware
+	r.Use(s.ConfigurationMiddleware())
+
+	// get the public key
+	r.Get("/conf", handlers.AppHandler(GetConfigurationHandler(s)).ServeHTTP)
+
+	// conf request in order to get service's pubkey so we can encrypt a secret
+	req1 := httptest.NewRequest("GET", "/conf", nil)
+	w1 := httptest.NewRecorder()
+	r.ServeHTTP(w1, req1)
+
+	var getConf = getConfResponse{}
+	if err := json.NewDecoder(w1.Result().Body).Decode(&getConf); err != nil {
+		t.Error("failed to decode config response", err)
+	}
+
+	// servicePubKey is the ed25519 key we will use for encrypting the config encryption key
+	servicePubKey, err := hex.DecodeString(getConf.PublicKey)
+	if err != nil {
+		t.Error("failed to decode config response", err)
+	}
+
+	// generate an ephemeral sender keypair
+	senderPubKey, senderPrivKey, err := box.GenerateKey(rand.Reader)
+	if err != nil {
+		t.Error("failed to create sender keypair", err)
+	}
+
+	var secretKey [32]byte
+	if _, err := io.ReadFull(rand.Reader, secretKey[:]); err != nil {
+		t.Error("failed to create random secret key", err)
+	}
+
+	// encrypt a secret
+	c, n, err := cryptography.EncryptMessage(secretKey, []byte(`{"version":"value"}`))
+	if err != nil {
+		t.Error("failed to encrypt our secrets", err)
+	}
+
+	// put the nonce at the start of our message
+	c = append(n[:], c...)
+
+	// now for the encryption of the secret for transport
+	var nonce [24]byte
+	if _, err := io.ReadFull(rand.Reader, nonce[:]); err != nil {
+		t.Error("failed to generate our nonce", err)
+	}
+
+	var rpk [32]byte
+	copy(rpk[:], servicePubKey[:32])
+
+	var spk [32]byte
+	copy(spk[:], senderPrivKey[:32])
+
+	// encrypt the key which was used to encrypt secret
+	encrypted := box.Seal(nonce[:], secretKey[:], &nonce, &rpk, &spk)
+
+	// now put this in our configuration
+	s.secretMgr = &mockSecretManager{
+		result: c, // c is our encrypted config items
+		err:    nil,
+	}
+
+	// set the values
+	r.Patch("/conf", handlers.AppHandler(PatchConfigurationHandler(s)).ServeHTTP)
+
+	r.Get("/valid", func(w http.ResponseWriter, r *http.Request) {
+		if v, ok := r.Context().Value(appctx.VersionCTXKey).(string); ok && v == "value" {
+			secretsStored = true
+		}
+		if v, ok := r.Context().Value(appctx.CommitCTXKey).(string); ok && v == "value" {
+			confStored = true
+		}
+		fmt.Println(w.Write([]byte("ok")))
+	})
+
+	reqBody := configurationHandlerRequest(map[appctx.CTXKey]interface{}{
+		appctx.CommitCTXKey:                  "value",                                      // commit is a configuration pushed in
+		appctx.SecretsURICTXKey:              "secrets uri",                                // tell configuration to pull new secrets
+		appctx.PaymentsEncryptionKeyCTXKey:   base64.StdEncoding.EncodeToString(encrypted), // tell configuration to pull new secrets
```

Comment:
What's the threat model and motivation here that led to us using the nitro_shim to encrypt a blob and store it in an S3 bucket and then encrypt the symmetrical key for the payments service? I can think of a few ways that we could simplify this architecture, but I'm not certain of the requirements that led us to this construction or that I properly understand the design so I want to understand these things first before suggesting changes.

Reply:
Possibly we can figure a way to archive/eliminate existing s3 buckets